### PR TITLE
Add "reshuffle env" command

### PR DESCRIPTION
**Pull request by arielshaqed**
This implements the [envariables CLI command](https://www.notion.so/reshuffle/0bb8970a81b4463d93e25f8297d8540a?v=ad5cd43ab2514427838488b57e929cbf&p=aacedd93da3e4a0ca3f62c06ee0645d3). However it does *not* currently adhere to that spec, so _cannot_ close
that task.

Tested vs. my realm and also unit-tested.

### CODE REVIEW

File: cli/src/utils/cli-flags.ts [hunk 1]
```diff
@@ -29,7 +29,13 @@ export default {
         return val;
       }
       error(`Expected a formatted duration or ISO 8601 string format but received: ${val}`);
-      throw new Error('Should not get here');
+    },
+  }),
+  keyValue: flags.build({
+    parse(val) {
+      const kvMatch = val.match(/^([^=]+)=(.*)$/);
```
Comment (vogre) [inline cli/src/utils/cli-flags.ts]: Can you add a test for the edge case of `VARIABLE=` empty string? I guess the only way to pass multiline strings as variables is to put them in an environment variable first?

File: cli/src/commands/env.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      await this.list(appId);
+    }
+  }
+
+  protected async get(appId: string, variableNames: string[]) {
+    const { variables } = await this.lycanClient.getEnv(appId, variableNames);
+    const indexed = fromPairs(variables.map(({ variable, value }) => [variable, value]));
+    // Output in requested order.
+    for (const variable of variableNames) {
+      this.log(`${variable}=${indexed[variable]}`);
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply cli/src/commands/env.ts]: Yes, it will (and I was expecting that and accepted it).  No, it cannot (and I was _not_ expecting that, and I'm not sure I accept it).
Suggestions?  I _could_ fail with an error, but that is probably not very nice -- you cannot see all the other variables.  OTOH, any string might be a value.

File: cli/src/commands/env.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
+import Command from '../utils/command';
+import flags from '../utils/cli-flags';
+import { getEnvOrDie } from '../utils/helpers';
+import fromPairs from 'lodash.frompairs';
+
+function nonEmpty<T>(x: T[]) { return x.length > 0 }
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply cli/src/commands/env.ts]: Another word from you and it will be
```ts
import * as R from 'ramda';
const nonEmpty = R.pipe(R.isEmpty, R.not);
```
# ❗️

File: cli/src/utils/cli-flags.ts (hunk 1 above)
Comment (arielshaqed) [reply cli/src/utils/cli-flags.ts]: Adding an empty variable to the test.

Multiline variables are easy using The Power of Bash (the `$'` is just to escape the single quotes, nothing to do with the multiline string).  But actually trying it revealed a bug in the `keyValue` arg type, so pushing a fix.
```sh
flutterby:counter-class ariels$ node  ~/Dev/reshuffle/cli/bin/run env --set $'a=
Humpty Dumpty sat on a wall,
Humpty Dumpty had a great fall.
All the king\'s horses
And all the king\'s men
Couldn\'t put Humpty together again.'
flutterby:counter-class ariels$ node  ~/Dev/reshuffle/cli/bin/run env --get a
a=
Humpty Dumpty sat on a wall,
Humpty Dumpty had a great fall.
All the king's horses
And all the king's men
Couldn't put Humpty together again.
```

File: cli/src/commands/env.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (vogre) [reply cli/src/commands/env.ts]: Other commands like `list` have a `--format=json` flag which would be unambiguous when you really need it.

If only needed for debugging might be enough to make sure that values are shown when running with `DEBUG=*`.

File: cli/src/commands/env.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    `$ ${Command.cliBinName} env --list --name fluffy-teapot-76`,
+    `$ ${Command.cliBinName} env --get PGUSER --get PGPASSWORD`,
+    `$ PGPASSWORD=secret ${Command.cliBinName} env --set PGUSER=admin --set-from-env PGPASSWORD`,
+  ];
+
+  public static args = [];
+
+  public static flags = {
+    ...Command.flags,
+    name: flags.string({
```
Comment (vogre) [inline cli/src/commands/env.ts]: Should this be `--app-name` for consistency across commands?

File: cli/src/commands/env.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (vogre) [reply cli/src/commands/env.ts]: 🤐

File: app-testsuite/cypress/integration/full_spec.js [hunk 5]
```diff
@@ -10,7 +10,16 @@ describe('counter', () => {
...
     cy.visit('/');
-    cy.get('.secret').contains('shhh');
+    cy.get('.secret').should('have.text', 'shhhh');
+  });
+});
+
+describe('app secret', () => {
+  it('gets env var set from secrets API', () => {
+    cy.visit('/');
+    if (cy.get('.runningOn').invoke('text') === 'realm') {
```
Comment (vogre) [inline app-testsuite/cypress/integration/full_spec.js]: This won't work as cypress is not actually synchronous, I think [then](https://docs.cypress.io/api/commands/then.html#Syntax) should work. Cypress API is actually async, but not in the `async/await` way (I think this was a good source https://medium.com/@NicholasBoll/cypress-io-using-async-and-await-4034e9bab207).

File: cli/src/commands/env.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaeladda) [reply cli/src/commands/env.ts]: You can simply use escaping, for example with `${JSON.stringify(indexed[variable])}` to tell apart the string undefined from the value undefined, but it is a bit cumbersome

File: cli/src/commands/env.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+export default class Env extends Command {
+  public static description = 'manipulate environment of deployed app';
+
+  public static examples = [
+    `$ ${Command.cliBinName} env --list --name fluffy-teapot-76`,
+    `$ ${Command.cliBinName} env --get PGUSER --get PGPASSWORD`,
+    `$ PGPASSWORD=secret ${Command.cliBinName} env --set PGUSER=admin --set-from-env PGPASSWORD`,
+  ];
+
```
Comment (michaeladda) [inline cli/src/commands/env.ts]: consider using a hidden command here please, at least until the interface solidify

File: app-testsuite/src/run.ts [hunk 7]
```diff
@@ -291,6 +295,11 @@ class App {
     await spawn('node', [CLI, 'deploy'], {
       cwd: this.appDir,
       stdio: 'inherit',
+      env: { ...process.env, REACT_APP_RUNNING_ON: 'realm' },
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply app-testsuite/src/run.ts]: Doing nothing, because it assumes remote build will behave in a particular way -- one which would surprise me.  I'm not sure why a comment here would stick out (unfortunately).

File: cli/src/utils/cli-flags.ts [hunk 8]
```diff
@@ -29,7 +29,13 @@ export default {
         return val;
       }
       error(`Expected a formatted duration or ISO 8601 string format but received: ${val}`);
-      throw new Error('Should not get here');
+    },
+  }),
+  keyValue: flags.build({
+    parse(val) {
+      const kvMatch = /^([^=]+)=(.*)$/s.exec(val);
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply cli/src/utils/cli-flags.ts]: Thanks, I see it -- and I think it's a strange and somewhat dangerous implementation.  Also it allows passing env vars through.  I explicitly chose *not* to pass them like this -- but will know more once we decide the UX.
